File: 14. Learn HTTP Clients in Go/Ch 2: JSON/6. Unmarshal JSON/main_test.go
package main

import (
	"fmt"
	"reflect"
	"testing"
)

func Test(t *testing.T) {
	type testCase struct {
		url      string
		expected []Issue
	}

	runCases := []testCase{
		{
			"https://api.boot.dev/v1/courses_rest_api/learn-http/issues?limit=1",
			[]Issue{{Title: "Fix that one bug nobody understands", Estimate: 19}},
		},
		{
			"https://api.boot.dev/v1/courses_rest_api/learn-http/issues?limit=2",
			[]Issue{
				{Title: "Fix that one bug nobody understands", Estimate: 19},
				{Title: "Implement user authentication flow", Estimate: 6},
			},
		},
	}

	submitCases := append(runCases, []testCase{
		{"", nil},
	}...)

	testCases := runCases
	if withSubmit {
		testCases = submitCases
	}

	skipped := len(submitCases) - len(testCases)

	passCount := 0
	failCount := 0

	for _, test := range testCases {
		items, _ := getIssues(test.url)

		if !reflect.DeepEqual(items, test.expected) {
			failCount++
			t.Errorf(`---------------------------------
URL:		%v
Expecting:  %+v
Actual:     %+v
Fail
`, test.url, test.expected, items)
		} else {
			passCount++
			fmt.Printf(`---------------------------------
URL:		%v
Expecting:  %+v
Actual:     %+v
Pass
`, test.url, test.expected, items)
		}
	}

	fmt.Println("---------------------------------")
	if skipped > 0 {
		fmt.Printf("%d passed, %d failed, %d skipped\n", passCount, failCount, skipped)
	} else {
		fmt.Printf("%d passed, %d failed\n", passCount, failCount)
	}
}

// withSubmit is set at compile time depending
// on which button is used to run the tests
var withSubmit = true

File: 1. Learn to Code in Python/Ch 11: Sets/6. Set Subtraction/main_test.py
from main import *

run_cases = [
    ([1, 1, 1, 2, 2, 2, 3], [1, 2], {3}),
    ([1, 2, 2, 3, 4, 3, 4, 5, 6, 7, 8, 9, 9, 10], [1, 2, 2, 3, 4, 5, 6, 7, 8], {9, 10}),
]

submit_cases = run_cases + [
    ([], [], set()),
    ([1, 1, 1], [], {1}),
    ([1, 2, 3, 4, 5], [1, 2, 3, 4, 5], set()),
    ([1, 1, 2, 2, 3, 3], [1, 2, 3], set()),
    ([1, 2, 3, 4, 5], [1, 2, 3], {4, 5}),
    ([1, 2, 3, 4, 5], [1, 3, 5], {2, 4}),
]


def test(input1, input2, expected_output):
    print("---------------------------------")
    print("Inputs:")
    print(f" - first_ids:  {input1}")
    print(f" - second_ids: {input2}")
    result = find_missing_ids(input1, input2)
    print(f"Expecting: {expected_output}")
    print(f"Actual:    {result}")
    if result == expected_output:
        print("Pass")
        return True
    print("Fail")
    return False


def main():
    passed = 0
    failed = 0
    skipped = len(submit_cases) - len(test_cases)
    for test_case in test_cases:
        correct = test(*test_case)
        if correct:
            passed += 1
        else:
            failed += 1
    if failed == 0:
        print("============= PASS ==============")
    else:
        print("============= FAIL ==============")
    if skipped > 0:
        print(f"{passed} passed, {failed} failed, {skipped} skipped")
    else:
        print(f"{passed} passed, {failed} failed")


test_cases = submit_cases
if "__RUN__" in globals():
    test_cases = run_cases

main()

